### PR TITLE
[GLUTEN-2582][CH] Bug fix crash problem in array<struct>/map<key, struct>/element_at query

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#2582 #2641)

relay on pr: https://github.com/Kyligence/ClickHouse/pull/457

## How was this patch tested?
UT

(If this patch involves UI changes, please attach a screenshot; otherwise, remove this)

### CODE REVIEW

File: cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    }
+
+protected:
+    String getCHFunctionName(const substrait::Expression_ScalarFunction & /*substrait_func*/) const override
+    {
+        return "arrayElement";
+    }
+};
+
+class FunctionParserElementAt : public FunctionParserArrayElement
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp:82]: 最好能每个scalar function对应一个文件。
另外需要将SCALAR_FUNCTION中对应的kv entry删掉。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Parser/TypeParser.cpp [hunk 2]
```diff
@@ -223,13 +225,13 @@ DB::Block TypeParser::buildBlockFromNamedStruct(const substrait::NamedStruct & s
     std::list<std::string> field_names;
     for (int i = 0; i < struct_.names_size(); ++i)
         field_names.emplace_back(struct_.names(i));
-
+
```
Comment (lgbo-ustc) [inline cpp-ch/local-engine/Parser/TypeParser.cpp:228]: There are space here?

File: cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp (hunk 1 above)
Comment (taiyang-li) [reply cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp:82]: 趁此机会将涉及到的几个function的spark ut开启，看看能否pass

File: cpp-ch/local-engine/Parser/TypeParser.cpp (hunk 2 above)
Comment (KevinyhZou) [reply cpp-ch/local-engine/Parser/TypeParser.cpp]: fixed done.

File: cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp (hunk 1 above)
Comment (KevinyhZou) [reply cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.cpp]: arrayElement 相关的function 目前主要包括：get_array_item, element_at, get_map_value，目前都已经添加到UT中了。

File: cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+{
+
+class BaseFunctionParserArrayElement  : public FunctionParser
+{
+public:
+    explicit BaseFunctionParserArrayElement(SerializedPlanParser * plan_parser_) : FunctionParser(plan_parser_) {}
+    ~BaseFunctionParserArrayElement() override = default;
+
+    const ActionsDAG::Node * parse(const substrait::Expression_ScalarFunction & substrait_func, ActionsDAGPtr & actions_dag) const override
+    {
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/scalar_function_parser/arrayElement.h:41]: 最好能添加一些注释，说明最终返回的function node是如何组成的，可参考其他function parser

**Review by taiyang-li — CHANGES_REQUESTED** [review]

**Comment by baibaichen** [comment]
@KevinyhZou you need update clickhouse.version, the commit id is  `461c7a0bf12`

**Review by baibaichen — APPROVED** [review]
LGTM

**Pull request merged**